src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -105,7 +105,7 @@ public DynamicallyAccessedMembersTypeHierarchy (LinkContext context, MarkStep ma
 			// method to look for annotations. If the newly marked type derives from the one we're processing
 			// it will rely on the cache to know if it's annotated - so the record must be in the cache
 			// before it happens.
-			if (apply) {
+			if (apply || (annotation != DynamicallyAccessedMemberTypes.None && _context.IsOptimizationEnabled (CodeOptimizations.ApplyTypeHierarchyAnnotation, type))) {
```

Comment:
An alternative solution would be to set `apply` to true (if the optimization is disabled) on line 56. That would have the added benefit, that the caches would be "in-sync" with the behavior (the cache would store `apply=true` as well). Would also mean that finding the actual `GetType` call for the hierarchy would not do anything (since the cache would tell it that it's already applied) - unlike as-is where that would try to apply the annotation again.